<!DOCTYPE html>
<html>

<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>README.md</title>
  <link rel="stylesheet" href="https://stackedit.io/style.css" />
</head>

<body class="stackedit">
  <div class="stackedit__html"><h1 id="locatit">LocatIt</h1>
<p>LocatIt processes the Molecular Barcode (MBC) information of a HaloPlex<sup>HS</sup>, SureSelect XT HS, or SureSelect XT HS2 Illumina sequencing run. For HaloPlex<sup>HS</sup> and SureSelect XT HS analyses, LocatIt will tag read pairs in a SAM/BAM file with their MBC sequences and mark or merge MBC duplicates from that SAM/BAM file. For SureSelect XT HS2 analyses, LocatIt requires that the input bam file has already been annotated with the MBC sequences (using Trimmer and BWA-MEM with “-C” parameter, for example).<br>
<br></p>
<p><em>Note: This jar was compiled using Java version 8. Please make sure your Java Runtime Environment is at least at version 8 by running the command “java –version”.</em><br>
<br></p>
<h3 id="command-line-syntax">Command-line syntax</h3>
<ul>
<li>For SureSelect XT HS2  (assumes that the input file already has the correct MBC tags):</li>
</ul>
<pre><code>java -Xmx12G -jar locatit.jar [OPTIONS] input_bam_file_name
</code></pre>
<ul>
<li>For other applications:</li>
</ul>
<pre><code>java -Xmx12G -jar locatit.jar [OPTIONS] input_bam_file_name index2_fastq_file_1 index2_fastq_file_2
</code></pre>
<p><em><strong>Note: -Xmx12G is just an example, please adjust memory based on sequencing depth and size of the input file.</strong></em><br>
<br></p>
<h3 id="required-parameters">Required parameters:</h3>
<hr>

<table>
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>input_bam_file_name</code></td>
<td>name of the input BAM or SAM file.</td>
</tr>
<tr>
<td><code>index2_fastq_file_1..N</code></td>
<td>input fastq file(s) containing the MBC sequences for the read pairs in the input bam file (not applicable to SureSelect XT HS2 analyses). The fastq files must contain all reads that are in the S/BAM file, but may also contain reads that have been filtered out of the S/BAM during processing.</td>
</tr>
</tbody>
</table><hr>
<br>
<h3 id="options">Options:</h3>
<hr>

<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>-l &lt;covered.bed&gt;</code></td>
<td>Specify covered bed file downloaded from Agilent’s SureDesign. If specified, the properties file histogram reflects what is happening within and not within the covered region. For SureSelect designs, this option is identical to the -b covered.bed option described above.</td>
</tr>
<tr>
<td><code>‐b &lt;intervals.bed&gt;</code></td>
<td>Provide design files downloaded from Agilent’s SureDesign website. For Haloplex designs, use the amplicon bed file. For SureSelect designs, this option is identical to the -l.</td>
</tr>
<tr>
<td><code>‑o&nbsp;&lt;output_file_name&gt;</code></td>
<td>name of the file generated by LocatIt.</td>
</tr>
<tr>
<td><code>-v2Only</code></td>
<td>For SureSelect XT HS2, turn on the duplex consensus mode option. The input SAM/BAM file needs to already be annotated with the correct 3+3 MBC SAM tags. First, LocatIt performs single-strand consensus deduplication. When two complementary single-strand consensus sequences are present, they are used to generate a duplex consensus sequence. All single-strand consensus reads are marked as not passing the vendor quality test and should be filtered out prior to running any downstream analysis.</td>
</tr>
<tr>
<td><code>-i</code></td>
<td>Incremental; instead of having the list of amplicons, i.e. the list of all possible pair starts/stops, the program learns all the possible start/stop combinations as it is reading the data. This is the main option that switches between HaloPlex and SureSelect modes. LocatIt performs single-strand consensus deduplication for both SureSelect XT HS and XT HS2.</td>
</tr>
<tr>
<td><code>-v2</code></td>
<td><strong>Experimental, do not use.</strong> For SureSelect XT HS2, turns on the “hybrid” deduplication approach. The approach is the same as for -v2Only mode, but the single-strand consensus reads are not flagged as failing the vendor quality test. The resulting BAM file may cause downstream applications to under-value the duplicate consensus reads compared to the single-consensus reads.</td>
</tr>
<tr>
<td><code>-R</code></td>
<td>Output bam file that has duplicates(SAM flag 0x400), filtered reads(SAM flag 0x200), secondary(SAM flag 0x100) and supplementary(SAM flag 0x800) reads removed.</td>
</tr>
<tr>
<td><code>-d &lt;number&gt;</code></td>
<td>Barcode distance. If two populations of read (pairs) only differ by this number of bases, both populations will be processed as having the same barcode. Because several errors can be merged back into the main population, some barcodes that are more than this number of bases away from each other may end up being merged together. <br>Range is 0 to 5. Default value is 0.</td>
</tr>
<tr>
<td><code>-q &lt;number&gt;</code></td>
<td>Reads having barcodes with quality less than specified threshold will be filtered. <br>Range is 0 to 45. Default value is 25.</td>
</tr>
<tr>
<td><code>-Q &lt;number&gt;</code></td>
<td>Reads having any base that is lower than specified threshold will be filtered. <br>Range is 0 to 45. Default value is 0.</td>
</tr>
<tr>
<td><code>-m &lt;number&gt;</code></td>
<td>Parameter specifies minimum number of read pairs associated with a barcode (amplification level). Barcodes having less reads than specified threshold will be filtered. <br>Range is &gt;=1. Default value is 1.</td>
</tr>
<tr>
<td><code>-c &lt;number&gt;</code></td>
<td>Enable optical duplicate detection and offset between two duplicate clusters on the flow cell in order to consider them optical duplicates. The offset is specified as a radius in pixels. If the offset between two duplicate clusters is less than this threshold, then the clusters are considered to be optical duplicates. <br> The value expected is an integer &gt;= 0. For unpatterned Illumina flow cells, 100 is an appropriate value. For the patterned flow cells, 2500 is more appropriate. By default, optical duplicate detection is disabled.</td>
</tr>
<tr>
<td><code>‐U</code></td>
<td>unsorted BAM/SAM output ‐ Faster and requires less RAM.</td>
</tr>
<tr>
<td><code>‐S</code></td>
<td>sorted BAM/SAM output</td>
</tr>
<tr>
<td><code>-L</code></td>
<td>If the input SAM/BAM contains fewer reads than the index2 file(s) and they are in the same order, the option -L allows LocatIt to discard non-matching index2 reads as it processes the input file instead of buffering them in case the matching reads would show up later. This saves a lot of memory.</td>
</tr>
<tr>
<td><code>-C</code></td>
<td>Chimeric; For HaloPlex, this means that the pairs which match two different amplicons are kept. For SureSelect, it should be always on. If -i is specified it sets -C internally.</td>
</tr>
<tr>
<td><code>-r</code></td>
<td>To remove/mask read1 read2 common overlap, half on each side.</td>
</tr>
<tr>
<td><code>‐IB</code></td>
<td>input file is BAM (default is SAM)</td>
</tr>
<tr>
<td><code>-IS</code></td>
<td>input file is SAM (default is SAM)</td>
</tr>
<tr>
<td><code>‐OB</code></td>
<td>output file is BAM (default is BAM)</td>
</tr>
<tr>
<td><code>-OS</code></td>
<td>output file is SAM (default is BAM)</td>
</tr>
<tr>
<td><code>-P</code></td>
<td>Renames the SAM tags so that downstream pipelines expecting different conventions can be used.<br><em>Syntax:</em> -P[letter]:[new 2 letters tag],[next],…<br><em>Example:</em> <code>-PM:xm,Q:xq,q:nq,r:nr</code> <br>The tags that can be renamed are:<ul><li><code>M</code> for barcode sequence</li><li><code>Q</code> for barcode quality</li><li><code>q</code> for barcode consensus quality</li><li><code>r</code> for read consensus quality</li><li><code>N</code> for barcode name tag</li><li><code>a</code> for alt readnames tag</li><li><code>d</code> for readnames tag</li><li><code>1</code> for begin amplicon tag</li><li><code>2</code> for end amplicon tag</li></ul></td>
</tr>
<tr>
<td><code>-K</code></td>
<td>Keep the intermediate bam file. With this option a coordinate-sorted and barcode-annotated bam file will be written to the parent directory of input bam file. By default this is not applied and is only useful for debugging.</td>
</tr>
<tr>
<td><code>‐X &lt;temp_directory&gt;</code></td>
<td>Location of temporary bam files used to store overflow of matches. Temporary files will be deleted at program exit.</td>
</tr>
<tr>
<td><code>‐t &lt;temp_directory&gt;</code></td>
<td>Location of temporary bam files used to store overflow of matches. Temporary files are not deleted.</td>
</tr>
</tbody>
</table><hr>
<br>
<h3 id="usage-examples">Usage examples:</h3>
<ul>
<li>For SureSelect XT HS2</li>
</ul>
<pre><code>java -Xmx12G -jar locatit.jar \
     -S -v2Only -d 1 -m 3 -q 25 -Q 25 \
     -l Covered.bed -o test_output.bam \ 
     test_input.bam
</code></pre>
<ul>
<li>For SureSelect XT HS (and XT HS2 without duplex consensus)</li>
</ul>
<pre><code>java -Xmx12G -jar locatit.jar \ 
     -PM:xm,Q:xq,q:nQ,r:nR \
     -q 25 -m 1 -U -IS -OB -C -i -r \ 
     -l covered.bed -o test_OUTPUT Test_CCP.sam \
     CCP_1_AACGTGAT_L001_R2_001.fastq.gz \
     CCP_1_AACGTGAT_L001_R2_002.fastq.gz
</code></pre>
<ul>
<li>For Halo</li>
</ul>
<pre><code>java -Xmx12G -jar locatit.jar \
     -q 25 -m 1 -U -IS -OB \
     -b ISCA2bf_extraG_001001398178390_AllTracks_amplicons.bed \
     -o test_OUTPUT \
     Test_ICCG_Panel.sam \
     ICCG-repl1_S1_L001_I1_001.fastq.gz
</code></pre>
<br>
<h4 id="tags-for-sureselect-xt-hs2">Tags For SureSelect XT HS2:</h4>
<p>LocatIt might insert some tags into SAM records and here are their meanings:</p>
<ul>
<li>fi:Z:<em>justifications for being filtered out/flagged as not passing vendor quality test</em><br>
e.g.&nbsp;<code>fi:Z:BAD_READ2_QUALITY;BELOW_MBC_NREADS_LIMIT;</code></li>
<li>XI:i:<em>number of duplicates collapsed for the single-strand consensus read</em><br>
e.g. <code>XI:i:2</code></li>
<li>XJ:i:<em>the number of duplicates collapsed for the single-strand consensus read on the complementary strand. (XT HS2 duplex consensus only.)</em><br>
e.g.  <code>XJ:i:1</code></li>
</ul>
</div>
</body>

</html>
