<?xml version="1.0" encoding="iso-8859-1"?>
<!DOCTYPE HTML>
<html>
<head>
 <script type="text/javascript" src="/site_media/jquery/js/jquery-1.6.1.min.js"></script>

 <style type="text/css">
  #formwrap {
   line-height: 2em;
   background: #eef;
   margin: 10px;
   padding: 10px;
   height: 500px;
  }
  body {
   font-family: "Lucida Grande", "Lucida Sans Unicode", Verdana, Arial, Helvetica, sans-serif;
   font-size: 14px;
  }
  .center { margin-left:auto; margin-right:auto; }
  .help {cursor:help; border-bottom: 1px dotted #A9A9A9}
 </style>

 <script>
  $(function() {

    var runTypes = {};
    $.ajaxSetup( {async:false} );
    $.get( '/rundb/api/v1/runtype/?format=json', function(data) {
      $.each( data.objects, function(intIndex,result) {
        var runType = result.runType;
        var selName = result.description.trim();
        // debar unsuported applications here, e.g. runType != "GENS"
        if( selName != "" ) {
          runTypes[runType] = selName;
        }
      });
    });

    $('#show_advanced').click(function() {
      var state = $(this).text();
      if( state == "Show advanced options" ) {
        state = "Hide advanced options";
        $('.advanced').show();
        // set show/hide based on application selected
        setOptionDefaults(TB_plan_json.runType,true);
      } else {
        state = "Show advanced options";
        $('.advanced').hide();
      }
      $(this).text(state);
      return false; // cancels redirection
    });

    function setOptionDefaults(runType,forceUpdate) {
      forceUpdate = forceUpdate || false;
      $("#libtypetxt").html(runTypes[runType]);
      var showchecks = 'NYYYY';
      if( runType.substr(0,4) == "AMPS" || runType == "MIXED") {
        showchecks = 'YNYNY';
      } else if( runType == "TAG_SEQUENCING" ) {
        showchecks = 'YNYNY';
      } else if( runType == "TARS_16S" || runType == "RNA" ) {
        showchecks = 'NNYNN';
      }
      $("#sampid").toggle(showchecks[0]=='Y' || !runType);
      // do not update for hidden advanced option - unless forced
      if( !forceUpdate && $('#show_advanced').text() == "Show advanced options" ) return;
      $("#tpadding").toggle(showchecks[1]=='Y');
      $("#uniquereads").toggle(showchecks[2]=='Y');
      $("#nondups").toggle(showchecks[3]=='Y');
      $("#cov_depth1").toggle(showchecks[4]=='Y');
      $("#cov_depth2").toggle(showchecks[4]=='Y');
      $("#cov_depth3").toggle(showchecks[4]=='Y');
    }

    // populate settings avaialble to the current Run Plan application type
    setOptionDefaults(TB_plan_json.runType);

  });
 </script>
</head>

<body>
 <form id="coverageAnalysis" name="coverageAnalysis">
  <div style="text-align:center">
   <h1>Torrent Coverage Analysis Plugin</h1>
   <h3>Genome and Targeted Re-sequencing Coverage Analysis</h3>
  </div>
   <table class="center" cellpadding="5px">
    <tr>
     <td><span class="help" title="The application (aka library, run or enrichment) type currently selected in the run. Available plugin options are dependent on the application.">Library Type:</span></td>
     <td><span id="libtypetxt">As Planned</span></tr>
    <tr id="sampid" style="display:none">
     <td><span class="help"
       title="Check if this AmpliSeq library employed sampleID tracking amplicons.">
       Sample Tracking:</span></td>
     <td><input type="checkbox" id="sampleid" name="sampleid" value="Yes"/></td> </tr>
   </table>
   <table class="center" cellpadding="5px">
    <tr>
     <td colspan=2 style="text-align:center"><a id="show_advanced" href="">Show advanced options</a></td>
    </tr>
    <tr id="tpadding" class="advanced" style="display:none">
     <td><span class="help"
       title="Specifiy the number of bases to expand each target at both 3' and 5' ends. This option is intended for use only with the TargetSeq Library Type but may also be specified for the Generic Sequencing application type.">
       Target Padding:</span></td>
     <td><input type="text" size=6 id="padtargets" name="padtargets" value=0></select></td> </tr>
    <tr id="uniquereads" class="advanced" style="display:none">
     <td><span class="help"
       title="Check to analyze only reads mapped to a unique location in the reference. Non-uniquely mapped reads are those that could have equally well aligned to multiple locations and are typically mapped randomly to one.">
       Uniquely Mapped Reads:</span></td>
     <td><input type="checkbox" id="uniquemaps" name="uniquemaps" value="Yes"/></td> </tr>
    <tr id="nondups" class="advanced" style="display:none">
     <td><span class="help"
       title="Check to analyze only reads that are not demarked as duplicate reads. This only applies to reads aligned with the Mark Duplicates option checked, which is not typcial for reads aligned by the Ion Torrent analysis pipeline.">
       Non-duplicate Reads:</span></td>
     <td><input type="checkbox" id="nonduplicates" name="nonduplicates" value="Yes" checked="checked"/></td> </tr>
    <tr id="alignlen" class="advanced" style="display:none">
     <td><span class="help"
       title="Specifiy the minimum aligned length a that read must have to be included in analysis.">
       Minimum Aligned Length:</span></td>
     <td><input type="text" size=6 id="minalignlen" name="minalignlen" value=0></select></td> </tr>
    <tr id="mapq" class="advanced" style="display:none">
     <td><span class="help"
       title="Specifiy the minimum mapping quality (MAPQ) that a read must have to be included in analysis. Using value of 1 (or higher) is equivalent to checking the 'Uniquely Mapped Reads' option.">
       Minimum Mapping Quality:</span></td>
     <td><input type="text" size=6 id="minmapqual" name="minmapqual" value=0></select></td> </tr>
    <tr id="cov_depth1" class="advanced" style="display:none">
     <td><span class="help"
       title="Specify the first tier coverage depth at which percentage of target coverage is reported. This value must be at least 2 since coverage at 1x read depth will always be output. The default value of 20 means that the percentage of targets, total target bases and/or individual target bases with at least 20 reads is reported.">
       Tier 1 Coverage Depth:</span></td>
     <td><input type="text" size=6 id="covdepth1" name="covdepth1" value=20></select></td> </tr>
    <tr id="cov_depth2" class="advanced" style="display:none">
     <td><span class="help"
       title="Specify the second tier coverage depth at which percentage of target coverage is reported. This value must be greater than the value used for first tier coverage. The default value of 100 means that the percentage of targets, total target bases and/or individual target bases with at least 100 reads is reported.">
       Tier 2 Coverage Depth:</span></td>
     <td><input type="text" size=6 id="covdepth2" name="covdepth2" value=100></select></td> </tr>
    <tr id="cov_depth3" class="advanced" style="display:none">
     <td><span class="help"
       title="Specify the third tier coverage depth at which percentage of target coverage is reported. This value must be greater than the value used for second tier coverage. The default value of 500 means that the percentage of targets, total target bases and/or individual target bases with at least 500 reads is reported.">
       Tier 3 Coverage Depth:</span></td>
     <td><input type="text" size=6 id="covdepth3" name="covdepth3" value=500></select></td> </tr>
   </table>
   <input type="hidden" id="launch_mode" name="launch_mode" value="Autostart with plan plugin configuration"/>
   <br/>
   <div align="left">
    <h3>Description and Usage Notes</h3>
    <p>
This plugin generates statistics, downloadable data files and interactive visualization of coverage over
targeted regions of the reference genome.
The reports produced by this plugin reflects the application selected in the plan and will also depend on library type
for specific barcodes where the application has mixed sample sources (DNA and RNA).
Reports mostly differ with respect to whether coverage is targeted and
includes base reads.</p>
    <p>
The SampleID Tracking option is recommended if your AmpliSeq panel was 'spiked' with SampleID sample tracking targets.
This will inform the analysis to produce an extra statistic for reads mapped to these targets so that
the true level of 'off-target' reads is accurately represented. This option is only available for AmpliSeq applications
and typically only effective for reads aligned to the human hg19 or GRCh38 genome reference.</p>
    <p>
Advanced options are available upon clicking on the Show advanced options link. Some options will only be available
depending on the application selected in the plan wizard. Refer to the tooltip help by hovering the mouse over the option text to
see a detailed description of an individual option.</p>
   </div>
 </form>
</body>
</html>

